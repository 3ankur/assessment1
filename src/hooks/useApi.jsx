import React  from "react";

export const apiStates = {
    'LOADING': 'LOADING',
    'SUCCESS': 'SUCCESS',
    'ERROR': 'ERROR',
};

export const useApi = (url)=>{
    const [data,setData] = React.useState({
        state:apiStates.LOADING,
        error:'',
        data:[]
    });

    const setPartData = (partData)=> setData({...data,...partData});

    React.useEffect(()=>{
        setPartData({
            state: apiStates.LOADING
        })

        fetch(url)
        .then(response=>response.json())
        .then(res=>{
            console.log(res);
            setPartData({
                state: apiStates.SUCCESS,
                data:res.zones
            })
        })
        .catch((err)=>{
            setPartData({
                state: apiStates.ERROR,
                error: 'Failed'
            })
        })

      
    },[])
    return data;
}