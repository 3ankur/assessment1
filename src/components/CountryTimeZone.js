import React, { useState } from "react";
import { apiStates, useApi } from "../hooks/useApi.jsx";
import TimeZoneDetails from "./timeZoneDetails";
const API_END_POINT = 'http://api.timezonedb.com/';
let firstLoad = false;
const CountryTimeZone = () =>{
   
    const {data,error,state} = useApi(`${API_END_POINT}v2.1/list-time-zone?key=XWSLLPX5RMIZ&format=json`);
    const [activeTimeZone,setTimeZone] = useState(null);
    const [isLoading, setLoading] = useState( firstLoad ? false : true);

    function setNewTimeZone(newZone){
        setLoading(true);
       fetch(`${API_END_POINT}v2/get-time-zone?key=XWSLLPX5RMIZ&format=json&by=zone&zone=${newZone.zoneName}`)
       .then((response) => response.json())
       .then((data) => {
           if (data?.status === 'OK') {
               setLoading(false);
               setTimeZone(data);
           }
       });
    }

    switch(state){
        case 'ERROR': 
        return <div>{error}</div>
        case 'SUCCESS' : 
            return (
                <div className="timezone" data-testid="timezone">
                    <div className="wrapper">
        
                        {
                            data.length ? (
                                <ul className="timezone__list" data-testid="list">
                                {
                                       data.map((zone) => {
                                        return(
                                            <li className={activeTimeZone ? (activeTimeZone.zoneName === zone.zoneName ? 'active' : '') : '' } key={zone.zoneName}  onClick={()=>  setNewTimeZone(zone)}>
                                            <div>{zone.zoneName}</div>
                                        </li>
                                        )
                                    })
                                }
                            </ul>
                            ) : null
                        }
                        {                   
                            !isLoading ?  ( <TimeZoneDetails   activeTimeZone={activeTimeZone}/>) : (firstLoad ) && 'loading.. ' 
                        }
        
                    </div>
        
                </div>
            )
            default:
                return <p>loading..</p>;
    }
}

export default CountryTimeZone;