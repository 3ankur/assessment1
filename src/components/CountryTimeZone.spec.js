
import React from "react";
import renderer from "react-test-renderer";
import { useApi } from "../hooks/useApi.jsx";
import CountryTimeZone from "./CountryTimeZone";
import { render, waitFor, screen } from "@testing-library/react"

jest.mock('../hooks/useApi.jsx', () => ({
    useApi: () => ({
        state: 'SUCCESS',
        error: '',
        data: [{
            'countryCode': "AD",
            'countryName': "Andorra",
            'gmtOffset': 3600,
            'timestamp': 1612780836,
            'zoneName': "Europe/Andorra"
        }]
    })
}));

describe.only("List Timezone snapshot", () => {
    test('loading renders correctly', () => {
        const { getByTestId } = render(<CountryTimeZone />)
        expect(getByTestId('list')).toBeDefined();
        //for snapshot
        const tree  = renderer.create(<CountryTimeZone />).toJSON();
        console.log(tree)
        expect(tree).toMatchSnapshot();
    })
})