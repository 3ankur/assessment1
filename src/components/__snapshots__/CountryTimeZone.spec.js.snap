// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List Timezone snapshot loading renders correctly 1`] = `
<div
  className="timezone"
  data-testid="timezone"
>
  <div
    className="wrapper"
  >
    <ul
      className="timezone__list"
      data-testid="list"
    >
      <li
        className=""
        onClick={[Function]}
      >
        <div>
          Europe/Andorra
        </div>
      </li>
    </ul>
  </div>
</div>
`;
